From e7ad7e22160682245c5b39f207ebe3812d587ee9 Mon Sep 17 00:00:00 2001
From: Hongyong Huang <hongyong@qti.qualcomm.com>
Date: Wed, 20 Aug 2025 17:40:31 +0800
Subject: [PATCH] waylandsink: support gbm protocol and Q08C/Q10C.

- Support the wayland gbm_buffer_backend protocol. SHM formats, DMABuf
  formats and GBM formats are exposed differently in caps: the GBM
  formats are flagged with GST_CAPS_FEATURE_MEMORY_GBM. No buffer pool
  is proposed for GBM buffers, it is the upstream element responsibility
  to provide with such buffers.
- Add Q08C/Q10C for wayland to replace UBWC function before. NV12_Q08C
  is equal to NV12+UBWC, NV12_Q10LE32C is equal to
  NV12_10LE32+UBWC+10bit. P010_10LE is reserved.
- Fix rotate dark result issue when fullscreen=true.

Upstream-Status: Inappropriate [Qualcomm specific]

Signed-off-by: Hongyong Huang <hongyong@qti.qualcomm.com>
---
 ext/wayland/gstwaylandsink.c             |  36 ++++-
 gst-libs/gst/wayland/gstwldisplay.c      |  89 +++++++++++
 gst-libs/gst/wayland/gstwldisplay.h      |  13 ++
 gst-libs/gst/wayland/gstwlgbmallocator.c | 183 +++++++++++++++++++++++
 gst-libs/gst/wayland/gstwlgbmallocator.h |  42 ++++++
 gst-libs/gst/wayland/gstwlvideoformat.c  |  95 ++++++++----
 gst-libs/gst/wayland/gstwlvideoformat.h  |   9 ++
 gst-libs/gst/wayland/meson.build         |  11 +-
 gst-libs/gst/wayland/wayland.h           |   1 +
 9 files changed, 442 insertions(+), 37 deletions(-)
 create mode 100644 gst-libs/gst/wayland/gstwlgbmallocator.c
 create mode 100644 gst-libs/gst/wayland/gstwlgbmallocator.h

diff --git a/ext/wayland/gstwaylandsink.c b/ext/wayland/gstwaylandsink.c
index b3ca56d..ff02c9b 100644
--- a/ext/wayland/gstwaylandsink.c
+++ b/ext/wayland/gstwaylandsink.c
@@ -73,14 +73,16 @@ GST_DEBUG_CATEGORY (gstwayland_debug);
 
 #define WL_VIDEO_FORMATS \
     "{ BGRx, BGRA, RGBx, xBGR, xRGB, RGBA, ABGR, ARGB, RGB, BGR, " \
-    "RGB16, BGR16, YUY2, YVYU, UYVY, AYUV, NV12, NV21, NV16, NV61, " \
-    "YUV9, YVU9, Y41B, I420, YV12, Y42B, v308, NV12_Q08C }"
+    "RGB16, BGR16, YUY2, YVYU, UYVY, AYUV, NV12, P010_10LE, NV21, NV16, NV61, " \
+    "YUV9, YVU9, Y41B, I420, YV12, Y42B, v308, NV12_Q08C, NV12_Q10LE32C }"
 
 static GstStaticPadTemplate sink_template = GST_STATIC_PAD_TEMPLATE ("sink",
     GST_PAD_SINK,
     GST_PAD_ALWAYS,
     GST_STATIC_CAPS (GST_VIDEO_CAPS_MAKE (WL_VIDEO_FORMATS) ";"
         GST_VIDEO_CAPS_MAKE_WITH_FEATURES (GST_CAPS_FEATURE_MEMORY_DMABUF,
+            WL_VIDEO_FORMATS) ";"
+        GST_VIDEO_CAPS_MAKE_WITH_FEATURES (GST_CAPS_FEATURE_MEMORY_GBM,
             WL_VIDEO_FORMATS))
     );
 
@@ -656,7 +658,7 @@ gst_wayland_sink_get_caps (GstBaseSink * bsink, GstCaps * filter)
 
   if (self->display) {
     GValue shm_list = G_VALUE_INIT, dmabuf_list = G_VALUE_INIT;
-    GValue value = G_VALUE_INIT;
+    GValue gbm_list = G_VALUE_INIT, value = G_VALUE_INIT;
     GArray *formats;
     GArray *modifiers;
     gint i;
@@ -666,6 +668,7 @@ gst_wayland_sink_get_caps (GstBaseSink * bsink, GstCaps * filter)
 
     g_value_init (&shm_list, GST_TYPE_LIST);
     g_value_init (&dmabuf_list, GST_TYPE_LIST);
+    g_value_init (&gbm_list, GST_TYPE_LIST);
 
     /* Add corresponding shm formats */
     formats = gst_wl_display_get_shm_formats (self->display);
@@ -709,6 +712,17 @@ gst_wayland_sink_get_caps (GstBaseSink * bsink, GstCaps * filter)
     gst_structure_take_value (gst_caps_get_structure (caps, 1), "format",
         &dmabuf_list);
 
+    /* Add corresponding gbm formats */
+    formats = gst_wl_display_get_gbm_formats (self->display);
+    for (i = 0; i < formats->len; i++) {
+      g_value_init (&value, G_TYPE_STRING);
+      fmt = g_array_index (formats, uint32_t, i);
+      g_value_set_static_string (&value, gst_wl_gbm_format_to_string (fmt));
+      gst_value_list_append_and_take_value (&gbm_list, &value);
+    }
+    gst_structure_take_value (gst_caps_get_structure (caps, 2), "format",
+        &gbm_list);
+
     GST_DEBUG_OBJECT (self, "display caps: %" GST_PTR_FORMAT, caps);
   }
 
@@ -754,7 +768,8 @@ static gboolean
 gst_wayland_sink_set_caps (GstBaseSink * bsink, GstCaps * caps)
 {
   GstWaylandSink *self = GST_WAYLAND_SINK (bsink);;
-  gboolean use_dmabuf;
+  GstStructure *structure;
+  gboolean use_dmabuf, use_gbm;
   GstVideoFormat format;
 
   GST_DEBUG_OBJECT (self, "set caps %" GST_PTR_FORMAT, caps);
@@ -774,10 +789,16 @@ gst_wayland_sink_set_caps (GstBaseSink * bsink, GstCaps * caps)
   use_dmabuf = gst_caps_features_contains (gst_caps_get_features (caps, 0),
       GST_CAPS_FEATURE_MEMORY_DMABUF);
 
+  use_gbm = gst_caps_features_contains (gst_caps_get_features (caps, 0),
+      GST_CAPS_FEATURE_MEMORY_GBM);
+
   /* validate the format base on the memory type. */
   if (use_dmabuf) {
     if (!gst_wl_display_check_format_for_dmabuf (self->display, format))
       goto unsupported_format;
+  } else if (use_gbm) {
+    if (!gst_wl_display_check_format_for_gbm (self->display, format))
+      goto unsupported_format;
   } else if (!gst_wl_display_check_format_for_shm (self->display, format)) {
     goto unsupported_format;
   }
@@ -967,7 +988,8 @@ gst_wayland_sink_show_frame (GstVideoSink * vsink, GstBuffer * buffer)
       "display, creating it", buffer);
 
   format = GST_VIDEO_INFO_FORMAT (&self->video_info);
-  if (gst_wl_display_check_format_for_dmabuf (self->display, format)) {
+  if (self->use_dmabuf &&
+      gst_wl_display_check_format_for_dmabuf (self->display, format)) {
     guint i, nb_dmabuf = 0;
 
     for (i = 0; i < gst_buffer_n_memory (buffer); i++)
@@ -978,6 +1000,10 @@ gst_wayland_sink_show_frame (GstVideoSink * vsink, GstBuffer * buffer)
     if (nb_dmabuf && (nb_dmabuf == gst_buffer_n_memory (buffer)))
       wbuf = gst_wl_linux_dmabuf_construct_wl_buffer (buffer, self->display,
           &self->video_info);
+  } else if (gst_wl_display_check_format_for_gbm (self->display, format)) {
+    if (gst_buffer_n_memory (buffer) == 1 && gst_is_fd_memory (mem))
+      wbuf = gst_wl_gbm_memory_construct_wl_buffer (buffer, self->display,
+          &self->video_info);
   }
 
   if (!wbuf && gst_wl_display_check_format_for_shm (self->display, format)) {
diff --git a/gst-libs/gst/wayland/gstwldisplay.c b/gst-libs/gst/wayland/gstwldisplay.c
index f43713d..27645b6 100644
--- a/gst-libs/gst/wayland/gstwldisplay.c
+++ b/gst-libs/gst/wayland/gstwldisplay.c
@@ -28,8 +28,12 @@
 #include "linux-dmabuf-unstable-v1-client-protocol.h"
 #include "viewporter-client-protocol.h"
 #include "xdg-shell-client-protocol.h"
+#include "gbm-buffer-backend-client-protocol.h"
 
 #include <errno.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <gbm_priv.h>
 
 #define GST_CAT_DEFAULT gst_wl_display_debug
 GST_DEBUG_CATEGORY_STATIC (GST_CAT_DEFAULT);
@@ -53,11 +57,15 @@ typedef struct _GstWlDisplayPrivate
   GArray *shm_formats;
   GArray *dmabuf_formats;
   GArray *dmabuf_modifiers;
+  struct gbm_buffer_backend *gbm;
+  GArray *gbm_formats;
+  struct gbm_device *gbmdevice;
 
   /* private */
   gboolean own_display;
   GThread *thread;
   GstPoll *wl_fd_poll;
+  gint gbmfd;
 
   GMutex buffers_mutex;
   GHashTable *buffers;
@@ -87,6 +95,7 @@ gst_wl_display_init (GstWlDisplay * self)
   priv->shm_formats = g_array_new (FALSE, FALSE, sizeof (uint32_t));
   priv->dmabuf_formats = g_array_new (FALSE, FALSE, sizeof (uint32_t));
   priv->dmabuf_modifiers = g_array_new (FALSE, FALSE, sizeof (uint64_t));
+  priv->gbm_formats = g_array_new (FALSE, FALSE, sizeof (uint32_t));
   priv->wl_fd_poll = gst_poll_new (TRUE);
   priv->buffers = g_hash_table_new (g_direct_hash, g_direct_equal);
   g_mutex_init (&priv->buffers_mutex);
@@ -94,6 +103,7 @@ gst_wl_display_init (GstWlDisplay * self)
   gst_wl_linux_dmabuf_init_once ();
   gst_wl_shm_allocator_init_once ();
   gst_wl_videoformat_init_once ();
+  gst_wl_gbm_mallocator_init_once ();
 }
 
 static void
@@ -126,6 +136,7 @@ gst_wl_display_finalize (GObject * gobject)
   g_array_unref (priv->shm_formats);
   g_array_unref (priv->dmabuf_formats);
   g_array_unref (priv->dmabuf_modifiers);
+  g_array_unref (priv->gbm_formats);
   gst_poll_free (priv->wl_fd_poll);
   g_hash_table_unref (priv->buffers);
   g_mutex_clear (&priv->buffers_mutex);
@@ -139,6 +150,9 @@ gst_wl_display_finalize (GObject * gobject)
   if (priv->dmabuf)
     zwp_linux_dmabuf_v1_destroy (priv->dmabuf);
 
+  if (priv->gbm)
+    gbm_buffer_backend_destroy(priv->gbm);
+
   if (priv->xdg_wm_base)
     xdg_wm_base_destroy (priv->xdg_wm_base);
 
@@ -165,6 +179,11 @@ gst_wl_display_finalize (GObject * gobject)
     wl_display_disconnect (priv->display);
   }
 
+  if (priv->gbmfd) {
+    gbm_device_destroy (priv->gbmdevice);
+    close (priv->gbmfd);
+  }
+
   G_OBJECT_CLASS (gst_wl_display_parent_class)->finalize (gobject);
 }
 
@@ -261,6 +280,30 @@ gst_wl_display_check_format_for_dmabuf (GstWlDisplay * self,
   return FALSE;
 }
 
+gboolean
+gst_wl_display_check_format_for_gbm (GstWlDisplay * display,
+    GstVideoFormat format)
+{
+  GstWlDisplayPrivate *priv = gst_wl_display_get_instance_private (display);
+  GArray *formats;
+  guint i, gbm_fmt;
+
+  if (!priv->gbm)
+    return FALSE;
+
+  gbm_fmt = gst_video_format_to_wl_gbm_format (format);
+  if (gbm_fmt == (guint) - 1)
+    return FALSE;
+
+  formats = priv->gbm_formats;
+  for (i = 0; i < formats->len; i++) {
+    if (g_array_index (formats, uint32_t, i) == gbm_fmt)
+      return TRUE;
+  }
+
+  return FALSE;
+}
+
 static void
 handle_xdg_wm_base_ping (void *user_data, struct xdg_wm_base *xdg_wm_base,
     uint32_t serial)
@@ -302,6 +345,25 @@ registry_handle_global (void *data, struct wl_registry *registry,
     priv->dmabuf =
         wl_registry_bind (registry, id, &zwp_linux_dmabuf_v1_interface, 3);
     zwp_linux_dmabuf_v1_add_listener (priv->dmabuf, &dmabuf_listener, self);
+  } else if (g_strcmp0 (interface, "gbm_buffer_backend") == 0) {
+    priv->gbm =
+        wl_registry_bind (registry, id, &gbm_buffer_backend_interface, 1);
+
+    if ((priv->gbmfd = open ("/dev/dri/card0", O_RDWR)) < 0) {
+      priv->gbmfd = open ("/dev/fb0", O_RDWR);
+      g_return_if_fail (priv->gbmfd >= 0);
+    }
+
+    priv->gbmdevice = gbm_create_device (priv->gbmfd);
+    g_return_if_fail (priv->gbmdevice != NULL);
+
+    guint formats[] = {
+        GBM_FORMAT_NV12,
+        GBM_FORMAT_YCbCr_420_P010_VENUS,
+        GBM_FORMAT_YCbCr_420_TP10_UBWC,
+        GBM_FORMAT_YCbCr_420_SP_VENUS_UBWC,
+    };
+    g_array_append_vals (priv->gbm_formats, formats, G_N_ELEMENTS (formats));
   }
 }
 
@@ -431,6 +493,10 @@ gst_wl_display_new_existing (struct wl_display * display,
     g_warning ("Could not bind to zwp_linux_dmabuf_v1");
   }
 
+  if (!priv->gbm) {
+    g_warning ("Could not bind to gbm_buffer_backend");
+  }
+
   if (!priv->xdg_wm_base && !priv->fullscreen_shell) {
     /* If wl_surface and wl_display are passed via GstContext
      * xdg_shell and zwp_fullscreen_shell are not used.
@@ -597,6 +663,29 @@ gst_wl_display_has_own_display (GstWlDisplay * self)
   return priv->own_display;
 }
 
+GArray *gst_wl_display_get_gbm_formats (GstWlDisplay * self)
+{
+  GstWlDisplayPrivate *priv = gst_wl_display_get_instance_private (self);
+
+  return priv->gbm_formats;
+}
+
+struct gbm_buffer_backend *
+gst_wl_display_get_gbm_buffer_backend (GstWlDisplay * self)
+{
+  GstWlDisplayPrivate *priv = gst_wl_display_get_instance_private (self);
+
+  return priv->gbm;
+}
+
+struct gbm_device *
+gst_wl_display_get_gbm_device (GstWlDisplay * self)
+{
+  GstWlDisplayPrivate *priv = gst_wl_display_get_instance_private (self);
+
+  return priv->gbmdevice;
+}
+
 void
 gst_wl_display_force_release_buffers (GstWlDisplay * self)
 {
diff --git a/gst-libs/gst/wayland/gstwldisplay.h b/gst-libs/gst/wayland/gstwldisplay.h
index a1dc254..415bc04 100644
--- a/gst-libs/gst/wayland/gstwldisplay.h
+++ b/gst-libs/gst/wayland/gstwldisplay.h
@@ -61,6 +61,10 @@ GST_WL_API
 gboolean gst_wl_display_check_format_for_dmabuf (GstWlDisplay * self,
     GstVideoFormat format);
 
+GST_WL_API
+gboolean gst_wl_display_check_format_for_gbm (GstWlDisplay * self,
+    GstVideoFormat format);
+
 GST_WL_API
 struct wl_display *gst_wl_display_get_display (GstWlDisplay * self);
 
@@ -100,6 +104,15 @@ struct zwp_linux_dmabuf_v1 *gst_wl_display_get_dmabuf_v1 (GstWlDisplay * self);
 GST_WL_API
 gboolean gst_wl_display_has_own_display (GstWlDisplay * self);
 
+GST_WL_API
+GArray *gst_wl_display_get_gbm_formats (GstWlDisplay * self);
+
+GST_WL_API
+struct gbm_buffer_backend * gst_wl_display_get_gbm_buffer_backend (GstWlDisplay * self);
+
+GST_WL_API
+struct gbm_device * gst_wl_display_get_gbm_device (GstWlDisplay * self);
+
 GST_WL_API
 void gst_wl_display_force_release_buffers (GstWlDisplay * self);
 
diff --git a/gst-libs/gst/wayland/gstwlgbmallocator.c b/gst-libs/gst/wayland/gstwlgbmallocator.c
new file mode 100644
index 0000000..b368410
--- /dev/null
+++ b/gst-libs/gst/wayland/gstwlgbmallocator.c
@@ -0,0 +1,183 @@
+/* GStreamer Wayland video sink
+ *
+ * Copyright (C) 2016 STMicroelectronics SA
+ * Copyright (C) 2016 Fabien Dessenne <fabien.dessenne@st.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public
+ * License along with this library; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include "gstwlgbmallocator.h"
+#include "gbm-buffer-backend-client-protocol.h"
+
+#include <gst/allocators/gstfdmemory.h>
+#include <gbm_priv.h>
+
+GST_DEBUG_CATEGORY (gst_wl_gbm_debug);
+#define GST_CAT_DEFAULT gst_wl_gbm_debug
+
+void
+gst_wl_gbm_mallocator_init_once (void)
+{
+  static gsize _init = 0;
+
+  if (g_once_init_enter (&_init)) {
+    GST_DEBUG_CATEGORY_INIT (gst_wl_gbm_debug, "wl_gbm", 0,
+        "wl_gbm library");
+
+    g_once_init_leave (&_init, 1);
+  }
+}
+
+typedef struct
+{
+  GMutex lock;
+  struct wl_buffer *wbuffer;
+  GCond cond;
+} GbmBufferData;
+
+static void
+gbm_create_succeeded(void *data, struct gbm_buffer_params *params,
+    struct wl_buffer *wbuffer)
+{
+  GbmBufferData *gbmdata = (GbmBufferData *) data;
+
+  g_mutex_lock (&gbmdata->lock);
+
+  gbmdata->wbuffer = wbuffer;
+  gbm_buffer_params_destroy (params);
+  g_cond_signal (&gbmdata->cond);
+
+  g_mutex_unlock (&gbmdata->lock);
+}
+
+static void
+gbm_create_failed(void *data, struct gbm_buffer_params *params)
+{
+  GbmBufferData *gbmdata = (GbmBufferData *) data;
+
+  g_mutex_lock (&gbmdata->lock);
+
+  gbmdata->wbuffer = NULL;
+  gbm_buffer_params_destroy (params);
+  g_cond_signal (&gbmdata->cond);
+
+  g_mutex_unlock (&gbmdata->lock);
+}
+
+static const struct gbm_buffer_params_listener gbm_params_listener = {
+    gbm_create_succeeded,
+    gbm_create_failed
+};
+
+struct wl_buffer *
+gst_wl_gbm_memory_construct_wl_buffer (GstBuffer * buffer,
+    GstWlDisplay * display, const GstVideoInfo * info)
+{
+  gint fd, metafd, format, flags = 0;
+  guint width, height;
+  struct gbm_buffer_params *params;
+
+  struct gbm_import_fd_data bufinfo;
+  struct gbm_bo *bo;
+
+  GbmBufferData gbmdata;
+  gint64 timeout;
+
+  g_return_val_if_fail (gst_wl_display_check_format_for_gbm (display,
+      GST_VIDEO_INFO_FORMAT (info)), NULL);
+
+  fd = gst_fd_memory_get_fd(gst_buffer_peek_memory (buffer, 0));
+  format = gst_video_format_to_wl_gbm_format (GST_VIDEO_INFO_FORMAT (info));
+
+  width = GST_VIDEO_INFO_WIDTH (info);
+  height = GST_VIDEO_INFO_HEIGHT (info);
+
+  bufinfo.fd = fd;
+  bufinfo.width = width;
+  bufinfo.height = height;
+  bufinfo.format = format;
+
+  if (format == GBM_FORMAT_NV12) {
+    bufinfo.format = GBM_FORMAT_YCbCr_420_SP_VENUS;
+  } else if (format == GBM_FORMAT_YCbCr_420_SP_VENUS_UBWC) {
+    flags = GBM_BO_USAGE_UBWC_ALIGNED_QTI;
+  } else if (format == GBM_FORMAT_YCbCr_420_TP10_UBWC) {
+    flags = GBM_BO_USAGE_UBWC_ALIGNED_QTI;
+  }
+
+  bo = gbm_bo_import (gst_wl_display_get_gbm_device(display), GBM_BO_IMPORT_FD, &bufinfo, flags);
+  g_return_val_if_fail (bo != NULL, NULL);
+
+  gbm_perform (GBM_PERFORM_GET_METADATA_ION_FD, bo, &metafd);
+  g_return_val_if_fail (metafd != -1, NULL);
+
+  GST_DEBUG_OBJECT (display, "Creating wl_buffer from GBM of size %"
+      G_GSSIZE_FORMAT " fd %d, metafd %d, (%d x %d), format %s", info->size,
+      fd, metafd, width, height, gst_wl_gbm_format_to_string (format));
+
+  if (format == GBM_FORMAT_YCbCr_420_SP_VENUS_UBWC)
+    format = GBM_FORMAT_NV12;
+
+  if (GST_BUFFER_FLAG_IS_SET (buffer, GST_VIDEO_BUFFER_FLAG_INTERLACED)) {
+    GST_DEBUG_OBJECT (display, "interlaced buffer");
+    flags = GBM_BUFFER_PARAMS_FLAGS_INTERLACED;
+
+    if (!GST_BUFFER_FLAG_IS_SET (buffer, GST_VIDEO_BUFFER_FLAG_TFF)) {
+      GST_DEBUG_OBJECT (display, "with bottom field first");
+      flags |= GBM_BUFFER_PARAMS_FLAGS_BOTTOM_FIRST;
+    }
+  }
+
+  g_cond_init (&gbmdata.cond);
+  g_mutex_init (&gbmdata.lock);
+  g_mutex_lock (&gbmdata.lock);
+
+  params = gbm_buffer_backend_create_params (gst_wl_display_get_gbm_buffer_backend(display));
+
+  /* Request buffer creation */
+  gbm_buffer_params_add_listener (params, &gbm_params_listener, &gbmdata);
+  gbm_buffer_params_create (params, fd, metafd, width, height, format, flags);
+
+  /* Wait for the request answer */
+  wl_display_flush (gst_wl_display_get_display(display));
+
+  gbmdata.wbuffer = NULL;
+  timeout = g_get_monotonic_time () + G_TIME_SPAN_SECOND;
+
+  if (!g_cond_wait_until (&gbmdata.cond, &gbmdata.lock, timeout)) {
+    GST_ERROR_OBJECT (display, "GBM buffer creation timed out");
+    gbm_buffer_params_destroy (params);
+  }
+
+  if (!gbmdata.wbuffer) {
+    GST_ERROR_OBJECT (display, "Can't create GBM wl_buffer!");
+  } else {
+    GST_DEBUG_OBJECT (display, "Created GBM wl_buffer (%p): %dx%d, fmt=%.4s",
+        gbmdata.wbuffer, width, height, (char *) &format);
+  }
+
+  g_mutex_unlock (&gbmdata.lock);
+  g_mutex_clear (&gbmdata.lock);
+  g_cond_clear (&gbmdata.cond);
+
+  gbm_bo_destroy (bo);
+
+  return gbmdata.wbuffer;
+}
\ No newline at end of file
diff --git a/gst-libs/gst/wayland/gstwlgbmallocator.h b/gst-libs/gst/wayland/gstwlgbmallocator.h
new file mode 100644
index 0000000..378aa02
--- /dev/null
+++ b/gst-libs/gst/wayland/gstwlgbmallocator.h
@@ -0,0 +1,42 @@
+/* GStreamer Wayland video sink
+ *
+ * Copyright (C) 2016 STMicroelectronics SA
+ * Copyright (C) 2016 Fabien Dessenne <fabien.dessenne@st.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public
+ * License along with this library; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ */
+
+#ifndef __GST_WL_GBM_ALLOCATOR_H__
+#define __GST_WL_GBM_ALLOCATOR_H__
+
+#include "gstwldisplay.h"
+
+G_BEGIN_DECLS
+
+#ifndef GST_CAPS_FEATURE_MEMORY_GBM
+#define GST_CAPS_FEATURE_MEMORY_GBM "memory:GBM"
+#endif
+
+GST_WL_API
+void gst_wl_gbm_mallocator_init_once (void);
+
+GST_WL_API
+struct wl_buffer * gst_wl_gbm_memory_construct_wl_buffer (GstBuffer * buffer,
+    GstWlDisplay * display, const GstVideoInfo * info);
+
+G_END_DECLS
+
+#endif /* __GST_WL_GBM_ALLOCATOR_H__ */
\ No newline at end of file
diff --git a/gst-libs/gst/wayland/gstwlvideoformat.c b/gst-libs/gst/wayland/gstwlvideoformat.c
index d136a3f..3b3eaea 100644
--- a/gst-libs/gst/wayland/gstwlvideoformat.c
+++ b/gst-libs/gst/wayland/gstwlvideoformat.c
@@ -28,6 +28,8 @@
 #include "gstwlvideoformat.h"
 
 #include <drm_fourcc.h>
+#include <gbm.h>
+#include <gbm_priv.h>
 
 #define GST_CAT_DEFAULT gst_wl_videoformat_debug
 GST_DEBUG_CATEGORY_STATIC (GST_CAT_DEFAULT);
@@ -49,40 +51,43 @@ typedef struct
 {
   enum wl_shm_format wl_shm_format;
   guint dma_format;
+  guint gbm_format;
   guint64 dma_modifier;
   GstVideoFormat gst_format;
 } wl_VideoFormat;
 
 static const wl_VideoFormat wl_formats[] = {
-  {WL_SHM_FORMAT_XRGB8888, DRM_FORMAT_XRGB8888, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_BGRx},
-  {WL_SHM_FORMAT_ARGB8888, DRM_FORMAT_ARGB8888, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_BGRA},
-  {WL_SHM_FORMAT_XBGR8888, DRM_FORMAT_XBGR8888, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_RGBx},
-  {WL_SHM_FORMAT_RGBX8888, DRM_FORMAT_RGBX8888, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_xBGR},
-  {WL_SHM_FORMAT_BGRX8888, DRM_FORMAT_BGRX8888, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_xRGB},
-  {WL_SHM_FORMAT_ABGR8888, DRM_FORMAT_ABGR8888, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_RGBA},
-  {WL_SHM_FORMAT_RGBA8888, DRM_FORMAT_RGBA8888, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_ABGR},
-  {WL_SHM_FORMAT_BGRA8888, DRM_FORMAT_BGRA8888, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_ARGB},
-  {WL_SHM_FORMAT_RGB888, DRM_FORMAT_RGB888, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_BGR},
-  {WL_SHM_FORMAT_BGR888, DRM_FORMAT_BGR888, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_RGB},
-  {WL_SHM_FORMAT_RGB565, DRM_FORMAT_RGB565, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_RGB16},
-  {WL_SHM_FORMAT_BGR565, DRM_FORMAT_BGR565, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_BGR16},
-
-  {WL_SHM_FORMAT_YUYV, DRM_FORMAT_YUYV, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_YUY2},
-  {WL_SHM_FORMAT_YVYU, DRM_FORMAT_YVYU, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_YVYU},
-  {WL_SHM_FORMAT_UYVY, DRM_FORMAT_UYVY, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_UYVY},
-  {WL_SHM_FORMAT_AYUV, DRM_FORMAT_AYUV, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_AYUV},
-  {WL_SHM_FORMAT_NV12, DRM_FORMAT_NV12, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_NV12},
-  {WL_SHM_FORMAT_NV21, DRM_FORMAT_NV21, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_NV21},
-  {WL_SHM_FORMAT_NV16, DRM_FORMAT_NV16, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_NV16},
-  {WL_SHM_FORMAT_NV61, DRM_FORMAT_NV61, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_NV61},
-  {WL_SHM_FORMAT_YUV410, DRM_FORMAT_YUV410, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_YUV9},
-  {WL_SHM_FORMAT_YVU410, DRM_FORMAT_YVU410, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_YVU9},
-  {WL_SHM_FORMAT_YUV411, DRM_FORMAT_YUV411, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_Y41B},
-  {WL_SHM_FORMAT_YUV420, DRM_FORMAT_YUV420, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_I420},
-  {WL_SHM_FORMAT_YVU420, DRM_FORMAT_YVU420, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_YV12},
-  {WL_SHM_FORMAT_YUV422, DRM_FORMAT_YUV422, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_Y42B},
-  {WL_SHM_FORMAT_YUV444, DRM_FORMAT_YUV444, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_v308},
-  {WL_SHM_FORMAT_NV12, DRM_FORMAT_NV12, DRM_FORMAT_MOD_QCOM_COMPRESSED, GST_VIDEO_FORMAT_NV12_Q08C},
+  {WL_SHM_FORMAT_XRGB8888, DRM_FORMAT_XRGB8888, GBM_FORMAT_XRGB8888, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_BGRx},
+  {WL_SHM_FORMAT_ARGB8888, DRM_FORMAT_ARGB8888, GBM_FORMAT_ARGB8888, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_BGRA},
+  {WL_SHM_FORMAT_XBGR8888, DRM_FORMAT_XBGR8888, GBM_FORMAT_XBGR8888, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_RGBx},
+  {WL_SHM_FORMAT_RGBX8888, DRM_FORMAT_RGBX8888, GBM_FORMAT_RGBX8888, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_xBGR},
+  {WL_SHM_FORMAT_BGRX8888, DRM_FORMAT_BGRX8888, GBM_FORMAT_BGRX8888, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_xRGB},
+  {WL_SHM_FORMAT_ABGR8888, DRM_FORMAT_ABGR8888, GBM_FORMAT_ABGR8888, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_RGBA},
+  {WL_SHM_FORMAT_RGBA8888, DRM_FORMAT_RGBA8888, GBM_FORMAT_RGBA8888, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_ABGR},
+  {WL_SHM_FORMAT_BGRA8888, DRM_FORMAT_BGRA8888, GBM_FORMAT_BGRA8888, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_ARGB},
+  {WL_SHM_FORMAT_RGB888, DRM_FORMAT_RGB888, GBM_FORMAT_BGR888, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_RGB},
+  {WL_SHM_FORMAT_BGR888, DRM_FORMAT_BGR888, GBM_FORMAT_RGB888, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_BGR},
+  {WL_SHM_FORMAT_RGB565, DRM_FORMAT_RGB565, GBM_FORMAT_BGR565, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_RGB16},
+  {WL_SHM_FORMAT_BGR565, DRM_FORMAT_BGR565, GBM_FORMAT_RGB565, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_BGR16},
+
+  {WL_SHM_FORMAT_YUYV, DRM_FORMAT_YUYV, GBM_FORMAT_YUYV, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_YUY2},
+  {WL_SHM_FORMAT_YVYU, DRM_FORMAT_YVYU, GBM_FORMAT_YVYU, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_YVYU},
+  {WL_SHM_FORMAT_UYVY, DRM_FORMAT_UYVY, GBM_FORMAT_UYVY, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_UYVY},
+  {WL_SHM_FORMAT_AYUV, DRM_FORMAT_AYUV, GBM_FORMAT_AYUV, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_AYUV},
+  {WL_SHM_FORMAT_NV12, DRM_FORMAT_NV12, GBM_FORMAT_NV12, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_NV12},
+  {WL_SHM_FORMAT_NV21, DRM_FORMAT_NV21, GBM_FORMAT_NV21, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_NV21},
+  {WL_SHM_FORMAT_NV16, DRM_FORMAT_NV16, GBM_FORMAT_NV16, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_NV16},
+  {WL_SHM_FORMAT_NV61, DRM_FORMAT_NV61, GBM_FORMAT_NV61, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_NV61},
+  {WL_SHM_FORMAT_YUV410, DRM_FORMAT_YUV410, GBM_FORMAT_YUV410, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_YUV9},
+  {WL_SHM_FORMAT_YVU410, DRM_FORMAT_YVU410, GBM_FORMAT_YVU410, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_YVU9},
+  {WL_SHM_FORMAT_YUV411, DRM_FORMAT_YUV411, GBM_FORMAT_YUV411, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_Y41B},
+  {WL_SHM_FORMAT_YUV420, DRM_FORMAT_YUV420, GBM_FORMAT_YUV420, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_I420},
+  {WL_SHM_FORMAT_YVU420, DRM_FORMAT_YVU420, GBM_FORMAT_YVU420, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_YV12},
+  {WL_SHM_FORMAT_YUV422, DRM_FORMAT_YUV422, GBM_FORMAT_YUV422, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_Y42B},
+  {WL_SHM_FORMAT_YUV444, DRM_FORMAT_YUV444, GBM_FORMAT_YUV444, DRM_FORMAT_MOD_LINEAR, GST_VIDEO_FORMAT_v308},
+  {WL_SHM_FORMAT_P010, DRM_FORMAT_P010, GBM_FORMAT_YCbCr_420_P010_VENUS, DRM_FORMAT_MOD_QCOM_COMPRESSED, GST_VIDEO_FORMAT_P010_10LE},
+  {WL_SHM_FORMAT_NV12, DRM_FORMAT_NV12, GBM_FORMAT_YCbCr_420_SP_VENUS_UBWC, DRM_FORMAT_MOD_QCOM_COMPRESSED, GST_VIDEO_FORMAT_NV12_Q08C},
+  {WL_SHM_FORMAT_NV12, DRM_FORMAT_NV12, GBM_FORMAT_YCbCr_420_TP10_UBWC, DRM_FORMAT_MOD_QCOM_COMPRESSED, GST_VIDEO_FORMAT_NV12_Q10LE32C},
 };
 
 enum wl_shm_format
@@ -111,6 +116,19 @@ gst_video_format_to_wl_dmabuf_format (GstVideoFormat format)
   return -1;
 }
 
+gint
+gst_video_format_to_wl_gbm_format (GstVideoFormat format)
+{
+  guint i;
+
+  for (i = 0; i < G_N_ELEMENTS (wl_formats); i++)
+    if (wl_formats[i].gst_format == format)
+      return wl_formats[i].gbm_format;
+
+  GST_WARNING ("wayland gbm video format not found");
+  return -1;
+}
+
 guint64
 gst_video_format_to_wl_dmabuf_modifier (GstVideoFormat format)
 {
@@ -148,6 +166,18 @@ gst_wl_dmabuf_format_to_video_format (guint wl_format)
   return GST_VIDEO_FORMAT_UNKNOWN;
 }
 
+GstVideoFormat
+gst_wl_gbm_format_to_video_format (guint wl_format)
+{
+  guint i;
+
+  for (i = 0; i < G_N_ELEMENTS (wl_formats); i++)
+    if (wl_formats[i].gbm_format == wl_format)
+      return wl_formats[i].gst_format;
+
+  return GST_VIDEO_FORMAT_UNKNOWN;
+}
+
 GstVideoFormat
 gst_wl_dmabuf_format_with_modifier_to_video_format (guint wl_format, guint64 modifier)
 {
@@ -174,3 +204,10 @@ gst_wl_dmabuf_format_to_string (guint wl_format, guint64 modifier)
   return gst_video_format_to_string
       (gst_wl_dmabuf_format_with_modifier_to_video_format (wl_format, modifier));
 }
+
+const gchar *
+gst_wl_gbm_format_to_string (guint wl_format)
+{
+  return gst_video_format_to_string
+      (gst_wl_gbm_format_to_video_format (wl_format));
+}
diff --git a/gst-libs/gst/wayland/gstwlvideoformat.h b/gst-libs/gst/wayland/gstwlvideoformat.h
index 86ad1a2..c4ce6ff 100644
--- a/gst-libs/gst/wayland/gstwlvideoformat.h
+++ b/gst-libs/gst/wayland/gstwlvideoformat.h
@@ -38,6 +38,9 @@ enum wl_shm_format gst_video_format_to_wl_shm_format (GstVideoFormat format);
 GST_WL_API
 gint gst_video_format_to_wl_dmabuf_format (GstVideoFormat format);
 
+GST_WL_API
+gint gst_video_format_to_wl_gbm_format (GstVideoFormat format);
+
 GST_WL_API
 guint64 gst_video_format_to_wl_dmabuf_modifier (GstVideoFormat format);
 
@@ -47,6 +50,9 @@ GstVideoFormat gst_wl_shm_format_to_video_format (enum wl_shm_format wl_format);
 GST_WL_API
 GstVideoFormat gst_wl_dmabuf_format_to_video_format (guint wl_format);
 
+GST_WL_API
+GstVideoFormat gst_wl_gbm_format_to_video_format (guint wl_format);
+
 GST_WL_API
 GstVideoFormat gst_wl_dmabuf_format_with_modifier_to_video_format (
     guint wl_format, guint64 modifier);
@@ -57,4 +63,7 @@ const gchar *gst_wl_shm_format_to_string (enum wl_shm_format wl_format);
 GST_WL_API
 const gchar *gst_wl_dmabuf_format_to_string (guint wl_format, guint64 modifier);
 
+GST_WL_API
+const gchar *gst_wl_gbm_format_to_string (guint wl_format);
+
 G_END_DECLS
diff --git a/gst-libs/gst/wayland/meson.build b/gst-libs/gst/wayland/meson.build
index 5eb4484..db50770 100644
--- a/gst-libs/gst/wayland/meson.build
+++ b/gst-libs/gst/wayland/meson.build
@@ -1,10 +1,11 @@
 wl_req = '>= 1.15'
 wl_client_dep = dependency('wayland-client', version: wl_req, required: get_option('wayland'))
 libdrm_dep = dependency('libdrm', version: '>= 2.4.55', required: get_option('wayland'))
+libgbm_dep = dependency('gbm', version: '>= 21.1.1', required:true)
 wl_protocol_dep = dependency('wayland-protocols', version: wl_req, required: get_option('wayland'))
 wl_scanner = find_program('wayland-scanner', required: get_option('wayland'))
 # Also used in ext/wayland
-use_wayland = wl_protocol_dep.found() and wl_client_dep.found() and wl_scanner.found() and libdrm_dep.found()
+use_wayland = wl_protocol_dep.found() and wl_client_dep.found() and wl_scanner.found() and libdrm_dep.found() and libgbm_dep.found()
 
 if use_wayland
   wl_sources = [
@@ -13,6 +14,7 @@ if use_wayland
       'gstwldisplay.c',
       'gstwllinuxdmabuf.c',
       'gstwlshmallocator.c',
+      'gstwlgbmallocator.c',
       'gstwlvideobufferpool.c',
       'gstwlvideoformat.c',
       'gstwlwindow.c',
@@ -25,6 +27,7 @@ if use_wayland
       'gstwldisplay.h',
       'gstwllinuxdmabuf.h',
       'gstwlshmallocator.h',
+      'gstwlgbmallocator.h',
       'gstwlvideobufferpool.h',
       'gstwlvideoformat.h',
       'gstwlwindow.h',
@@ -38,6 +41,8 @@ if use_wayland
       ['/stable/viewporter/viewporter.xml', 'viewporter-protocol.c', 'viewporter-client-protocol.h'],
       ['/unstable/linux-dmabuf/linux-dmabuf-unstable-v1.xml',
        'linux-dmabuf-unstable-v1-protocol.c', 'linux-dmabuf-unstable-v1-client-protocol.h'],
+      ['/stable/gbm-buffer-backend/gbm-buffer-backend.xml',
+       'gbm-buffer-backend-protocol.c', 'gbm-buffer-backend-client-protocol.h'],
       ['/unstable/fullscreen-shell/fullscreen-shell-unstable-v1.xml',
        'fullscreen-shell-unstable-v1-protocol.c', 'fullscreen-shell-unstable-v1-client-protocol.h'],
       ['/stable/xdg-shell/xdg-shell.xml', 'xdg-shell-protocol.c', 'xdg-shell-client-protocol.h'],
@@ -74,7 +79,7 @@ if use_wayland
     darwin_versions : osxversion,
     install : true,
     dependencies : [gst_dep, gstallocators_dep, gstvideo_dep, libdrm_dep,
-                    wl_client_dep]
+                    wl_client_dep, libgbm_dep]
   )
 
   pkg_name = 'gstreamer-wayland-1.0'
@@ -90,7 +95,7 @@ if use_wayland
   gstwayland_dep = declare_dependency(link_with : gstwayland,
     include_directories : [libsinc],
     dependencies : [gst_dep, gstallocators_dep, gstvideo_dep, libdrm_dep,
-                    wl_client_dep])
+                    wl_client_dep, libgbm_dep])
 
   install_headers(wl_headers, subdir: 'gstreamer-1.0/gst/wayland')
   meson.override_dependency(pkg_name, gstwayland_dep)
diff --git a/gst-libs/gst/wayland/wayland.h b/gst-libs/gst/wayland/wayland.h
index 7a45845..25cff61 100644
--- a/gst-libs/gst/wayland/wayland.h
+++ b/gst-libs/gst/wayland/wayland.h
@@ -34,6 +34,7 @@
 #include <gst/wayland/gstwldisplay.h>
 #include <gst/wayland/gstwllinuxdmabuf.h>
 #include <gst/wayland/gstwlshmallocator.h>
+#include <gst/wayland/gstwlgbmallocator.h>
 #include <gst/wayland/gstwlvideobufferpool.h>
 #include <gst/wayland/gstwlvideoformat.h>
 #include <gst/wayland/gstwlwindow.h>
-- 
2.25.1

